### PR TITLE
Move environment functionality to utils

### PR DESCRIPTION
**Pull request by mjcarroll**
Since multiple ignition libraries make use of setting/getting environment variables, it makes more sense to locate it here.

### CODE REVIEW

File: src/Environment.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+/////////////////////////////////////////////////
+bool ignition::utils::env(const std::string &_name,
+                          std::string &_value,
+                          bool _allowEmpty)
+{
+  std::string v;
+  bool valid = false;
+#ifdef _WIN32
+  const DWORD buffSize = 32767;
```
Comment (chapulina) [inline src/Environment.cc:28]: Windows is sad

```
[26.422s] C:\Jenkins\workspace\ign_utils-pr-win\ws\ign-utils\src\Environment.cc(28,15): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int [C:\Jenkins\workspace\ign_utils-pr-win\ws\build\ignition-utils1\src\ignition-utils1.vcxproj]
```

**Review by chapulina — APPROVED** [review]
LGTM with green CI

File: src/Environment.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  if (_allowEmpty)
+  {
+    // Reading environment variable with _allowEmpty is unsupported on Windows
+    return false;
+  }
+
+#else
+  const char *cvar = std::getenv(_name.c_str());
```
Comment (azeey) [inline src/Environment.cc:47]: Needs `#include <cstdlib>`

File: src/Environment.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+/////////////////////////////////////////////////
+bool ignition::utils::env(const std::string &_name,
+                          std::string &_value,
+                          bool _allowEmpty)
+{
+  std::string v;
+  bool valid = false;
+#ifdef _WIN32
+  const DWORD buffSize = 32767;
+  static char buffer[buffSize];
```
Comment (azeey) [inline src/Environment.cc:29]: Does the fact that we're using a static char buffer mean this is not thread safe?

File: src/Environment.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  if (GetEnvironmentVariable(_name.c_str(), buffer, buffSize))
```
Comment (azeey) [inline src/Environment.cc:30]: Out of curiosity, do you know why we're using `GetEnvironmentVariable`? Does `std::getenv` not work on windows?

I found an [article](https://blogs.msmvps.com/senthil/2009/10/13/when-what-you-set-is-not-what-you-get-setenvironmentvariable-and-getenv/) that talks about the differences.

Comment (mjcarroll) [reply src/Environment.cc]: I believe `std::getenv` is deprecated on Windows, but I attempted a fix with `getenv_s` in 2fb0b28

File: src/Environment.cc (hunk 3 above)
Comment (mjcarroll) [reply src/Environment.cc]: Removed in 2fb0b28

File: src/Environment.cc (hunk 2 above)
Comment (mjcarroll) [reply src/Environment.cc:48]: Added in 2fb0b28

**Review by azeey — APPROVED** [review]
LGTM. The only caveat is that I don't think `env` and `setenv` are thread safe since the underlying functions are not. I'm not referring to the issue of a global `char *` getting invalidated by multiple calls to `getenv` from different threads, but the data race that can occur when both `setenv` and `getenv` are called from different threads. The windows [documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-s-wgetenv-s?view=msvc-170) recommends synchronizing calls to these functions. [cppreference](https://en.cppreference.com/w/cpp/utility/program/getenv) for `getenv` says:


> This function is thread-safe (calling it from multiple threads does not introduce a data race) as long as no other function modifies the host environment. In particular, the POSIX functions setenv(), unsetenv(), and putenv() would introduce a data race if called without synchronization.

Perhaps we can just document this in `Environment.hh`.

**Comment by mjcarroll** [comment]
> Perhaps we can just document this in Environment.hh.

Added documentation.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1

**Pull request merged**